File: LeetCode/MathProbs/java/powerXn/powerXn.java
package LeetCode.MathProbs.java.powerXn;

enum testResult {
    PASS("\u001B[32mPASS\u001B[0m"),
    FAIL("\u001B[31mFAIL\u001B[0m");

    private final String value;

    testResult(String value) {
        this.value = value;
    }

    public String getValue() {
        return this.value;
    }
}

record powerXnRecord(double x, int n, double expected) {
}

class powerXn {
    public double myPow(double x, int n) {
        if (n < 0) {
            n = -n;
            x = 1 / x;
        }

        double pow = 1;

        while (n != 0) {
            if ((n & 1) != 0) {
                pow *= x;
            }
            x *= x;
            n >>>= 1; // Unsigned rightshift
        }

        return pow;
    }

    public static void main(String[] args) {
        powerXnRecord[] records = new powerXnRecord[] {
                new powerXnRecord(2.00000, 10, 1024.00000),
                new powerXnRecord(2.10000, 3, 9.26100),
                new powerXnRecord(2.000000, -2, 0.25000)
        };

        int i = 1;
        for (powerXnRecord record : records) {
            System.out.println("# Test case " + i++);
            powerXn.testResult(record);
            System.out.println("-".repeat(50));
        }
    }

    private static void testResult(powerXnRecord record) {
        String formattedX = String.format("%.5f", record.x());
        System.out.println("input:\tx: " + formattedX);

        System.out.println("\tn: " + record.n());
        String formattedExpected = String.format("%.5f", record.expected());
        System.out.println("expected: " + formattedExpected);

        double res = new powerXn().myPow(record.x(), record.n());
        String formattedRes = String.format("%.5f", res);

        System.out.println("result: " + formattedRes);

        System.out.println(
                formattedRes.equals(formattedExpected) ? testResult.PASS.getValue() : testResult.FAIL.getValue());
    }
}

File: LeetCode/MathProbs/javascript/powerXn.js
const Result = { PASS: "\x1b[92mPASS\x1b[0m", FAIL: "\x1b[91mFAIL\x1b[0m" };

class Record {
  constructor(x, n, expected) {
    this.x = x;
    this.n = n;
    this.expected = expected;
  }
}

class Solution {
  /**
   * @param {number} x
   * @param {number} n
   * @return {number}
   */
  myPow(x, n) {
    if (n < 0) {
      n = -n;
      x = 1 / x;
    }

    let pow = 1;

    while (n != 0) {
      if ((n & 1) != 0) {
        pow *= x;
      }
      x *= x;
      n >>>= 1;
    }

    return pow;
  }
}

function testSolution(record) {
  console.log(`input:\tx: ${record.x.toFixed(5)}`);
  console.log(`\tn: ${record.n}`);
  console.log(`expected: ${record.expected.toFixed(5)}`);

  const solution = new Solution();
  const result = solution.myPow(record.x, record.n);

  console.log(`result: ${result.toFixed(5)}`);
  console.log(
    result.toFixed(5) === record.expected.toFixed(5) ? Result.PASS : Result.FAIL,
  );
}

const records = [
  new Record(2.0, 10, 1024.0),
  new Record(2.1, 3, 9.261),
  new Record(2.0, -2, 0.25),
];

records.forEach((record, i) => {
  console.log(`# Test case ${i + 1}`);
  testSolution(record);
  console.log("-".repeat(50));
});
